### DISCUSSION TITLE
LwIP-based Ethernet libraries for W5x00 and ENC28J60 supporting many other boards

### DISCUSSION THREAD

**Discussion by khoih-prog** [khoih-prog/Ethernet_Generic#2] [Ideas]
Hi @jandrassy

As we discussed quite some time ago about implementing **LwIP-based Ethernet** library for W5x00 and ENC28J60 for other boards (SAMD, nRF52, megaAVR, RP2040, etc), besides what already done here

- **ESP8266** (OK for both W5x00 and ENC), 
- **STM32** (only for built-in LAN8742A and LAN8720), 
- **Portenta_H7** Vision-shield Ethernet
- Teensy 4.1 **QNEthernet**, we're waiting until now and still nothing new yet.

I'm thinking about doing something as follows, **if time permitted**

1. Add support to **ENC28J60** to this library, so we can have transparent support to both W5x00 and ENC similar to what they did in ESP8266
2. Write / port W5x00 as well as  ENC28J60 to integrate with LwIP, so that Async functions are supported.

Any suggestion, help and/or idea are very well appreciate.

Regards,

**Comment by JAndrassy** [comment]
there is very big difference between enc28j60 handling and w5x00. Wiznet chips have the TCP/IP stack in firmware, for enc28j60 the TCP/IP stack is in the UIPEthernet/EthernetENC library (the stack is the old uIP library, simpler predecessor of LwIP).

Both Wiznet chips and Atmel ENC chips can run in a 'MAC level' mode used for example by the esp8288 lwip_w5500 and lwip_enc28j60 libraries. The STM32, Teensy 4 and ESP32 native Ethernet are 'MAC level' too I think. And many WiFi modules including the WiFi101 ATWIN1500 support the 'MAC level' too. (some are very cheap)
It is simple to add a new network interface to LwIP for these network modules.

The STM32 Arduino Ethernet library uses the STM made library based on LwIP https://github.com/stm32duino/LwIP. But they bundled the network interface with LwIP and wrote the Ethernet library using this LwIP library. Which is the wrong way. The LwIP and the Client, Server and UDP should be in one library and it should support libraries with the network interface implementation like the esp8266 WiFi library does with the lwip_* libraries.

If I would implement an Arduino networking library based on LwIP, I would start fresh from LwIP source, but if you want to try it I would recommend you to start with the STM32 LwIP library and make it so it can accept the esp8266 lwip_* libraries (including the lwip_ppp one). The Arduino Client, Server and UDP implementation could be from the STM32 Ethernet library.

**Reply by khoih-prog** [reply]
As you have very clear picture of `The Good, Bad and the Ugly`, it's much better and faster if you can spend time to finish this project.

If not possible, I'll find some time to start.

**Comment by JAndrassy** [comment]
I see there already is STM32AsyncTCP library so it could use the enhanced  STM LwIP  with support for esp8266 lwip_* librarries

**Reply by khoih-prog** [reply]
> But they bundled the network interface with LwIP and wrote the Ethernet library using this LwIP library

I did have a look at [**STM32duino LwIP**](https://github.com/stm32duino/LwIP) long time ago trying to do something similar. Starting directly from original lwIP is possibly easier, IMO.

I still prefer to use lwIP library as **independent** library (or **integrated in the cores**), and we just rewrite the **hardware-dependent** code for different interface (Ethernet/WiFi) around that lwIP, certainly with configurable options.

It's much better if all the remaining cores (SAMD, nRF52, STM32, RP2040, etc) **optionally** integrate the lwIP the same way the **much better ESP32/ESP8266 cores** are doing. Even if the WiFi/Ethernet is not integrated with the MCU.

I love the way the ESP8266 `lwIP_*` libraries are designed and hopefully we can do something similar to them.

The WiFiNINA, WiFi101, etc. will be much better to be lwIP-based.
